### DISCUSSION TITLE
Is IDL's behaviour for "OR" operand OK for all operand types and sizes? --- And what should GDL do?

### DISCUSSION THREAD

**Discussion by GillesDuvert** [gnudatalanguage/gdl#1782] [Q&A]
We believe we have tested ALL possible cases of ALL operands ( +, -,  /, ... AND, OR, ... etc) on ALL IDL types and sizes.
The results match those of IDL (fortunately) except for OR (and its sibling XOR) in ONE case, and this case is problematic, hence this call for advice:

As of today, if A is a integer-type scalar, say, 33, and B is an integer-type array, say, indgen(10) we have:
```
IDL> a=33 & b=indgen(10) & print, A OR B , B OR A
      33      33      35      35      37      37      39      39      41      41
      33      33      35      35      37      37      39      39      41      41
```
and
```
GDL> a=33 & b=indgen(10) & print, A OR B , B OR A
      33      33      35      35      37      37      39      39      41      41
      33      33      35      35      37      37      39      39      41      41
```
**So, same results.**

However **if A or B is of FLOAT TYPE**, the operands are of float type, and the results differ between IDL and GDL:
```
IDL> a=33. & b=findgen(10) & print, A OR B , B OR A
      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000
      33.0000      1.00000      2.00000      3.00000      4.00000      5.00000      6.00000      7.00000      8.00000      9.00000

```
```
GDL> a=33. & b=indgen(10) & print, A OR B , B OR A
      33.0000      1.00000      2.00000      3.00000      4.00000      5.00000      6.00000      7.00000      8.00000      9.00000
      33.0000      1.00000      2.00000      3.00000      4.00000      5.00000      6.00000      7.00000      8.00000      9.00000

```
So, with IDL, (A OR B) is NOT (B OR A) for floats. It is (A OR B[0]) replicated N_ELEMENTS(B).  I would say that IDL is wrong in this case. What do you think?

(this is a niche case --- I wonder if there is code somewhere that falls in this special case, so I would tend to not change the GDL behaviour).

**Comment by alaingdl** [comment]
following https://www.nv5geospatialsoftware.com/docs/Bitwise_Operators.html for float : 

_For floating- point operands, returns the first operand if it is non- zero, or the 2nd operand otherwise._

furthermore : 
_ Before the operation, the second operand is converted to the same type as the first operand. _

**Reply by GillesDuvert** [reply]
@alaingdl this is not what happens for IDL. The first operand is 33.0, not [33,33,33,...] , so the return for 
`a=33. & b=indgen(10) & print, A OR B`
should be `33.0000 ` not 
`33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000`
This is not, also, what happens for GDL, of course. But GDL at least insures `(A OR B) == (B OR A)` .

**Reply by alaingdl** [reply]
I fully understand the result we have in IDL since we have scalar : array combination
(See section Expression Structure here https://www.nv5geospatialsoftware.com/docs/data_type_and_structure.html#constants_818058343_699867)

and I understand too that OR operator it is *not* commutative for *floats* ...
(then (A OR B) != (B OR A) for floats)

```
non commutative for floats
IDL> a=33. & b=10. & print, A or B , B or A  
      33.0000      10.0000
IDL> a=33. & b=[10.] & print, A or B , B or A  
      33.0000
      10.0000

Commutative with Integers
IDL> a=33 & b=10 & print, A or B , B or A  
      43      43

combination of arrays and scalar
IDL> a=33. & b=findgen(10) & print, A or B , B or A  
      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000      33.0000
      33.0000      1.00000      2.00000      3.00000      4.00000      5.00000      6.00000      7.00000      8.00000      9.00000
IDL> a=[33.] & b=findgen(10) & print, A or B , B or A  
      33.0000
      33.0000

```

Last interesting point is for 0. with floats : 
```
IDL> a=1. & b=10. & print, A or B , B or A  
      1.00000      10.0000
IDL> a=0. & b=10. & print, A or B , B or A  
      10.0000      10.0000
```

**Reply by GillesDuvert** [reply]
@alaingdl thanks for pointing to this documentation I :blush: was not aware of, and for the example "combination of arrays and scalars".
Still, textbooks say that Order in an "and" or "or" doesn't matter, whatever the 'type' of A or B.
Anyway, as this seems more an internal logic of IDL than real "math" logic, I guess I should try to reproduces IDL's behaviour, soon a PR.

**Comment by GillesDuvert** [comment]
solved
